### ISSUE TITLE
feat(language): train nano-LLM on Tiny Shakespeare

### ISSUE THREAD

**Issue by leandronsp** [leandronsp/aspirina#6]
## Overview

Implement the training loop for the nano-LLM: cross-entropy loss, Adam optimizer, and train on Tiny Shakespeare (~1MB). The model learns to predict the next token. This is where all the pieces come together and we see the loss going down.

## Problem Statement

We have all the components (tokenizer, embedding, attention, transformer blocks) but no way to train them end-to-end. The existing backprop in gates/ uses MSE loss with SGD, which works for binary gates but not for language modeling. We need cross-entropy loss (comparing probability distributions over vocabulary) and a better optimizer.

## Current State

- Transformer blocks stack and produce output (from issue #5)
- Tokenizer produces token IDs from text (from issue #2)
- Embedding converts IDs to vectors (from issue #3)
- `gates/neural_network.rs` has backprop but with sigmoid + MSE. Not reusable for this

## Requirements

- [ ] R-1: Implement cross-entropy loss function and its derivative
- [ ] R-2: Implement output projection: final layer norm -> linear projection to vocab_size -> softmax to get next-token probabilities
- [ ] R-3: Implement Adam optimizer (or SGD with momentum as simpler alternative) with configurable learning rate
- [ ] R-4: Implement the full training loop: tokenize corpus -> create batches of (input_sequence, target_sequence) -> forward pass -> loss -> backward pass -> update weights
- [ ] R-5: Target sequence is input shifted by one position (standard causal LM training)
- [ ] R-6: Log training loss per epoch/step. Loss should decrease over training
- [ ] R-7: Train on Tiny Shakespeare with ~500K-2M parameters on M1 16GB in minutes to hours

## Technical Approach

```
language/src/
  train.rs        # TrainingConfig, training loop, loss computation
  model.rs        # NanoLLM struct: ties tokenizer + embedding + transformer stack + output projection
```

Model forward pass:
```
tokens = tokenizer.encode(text)           # Vec<u32>
x = embedding.forward(tokens)            # (seq_len, embed_dim)
for block in transformer_blocks:
  x = block.forward(x)                   # (seq_len, embed_dim)
x = layer_norm(x)                        # (seq_len, embed_dim)
logits = x * embedding_matrix.transpose  # (seq_len, vocab_size) — weight tying
probs = softmax(logits)                  # (seq_len, vocab_size)
loss = cross_entropy(probs, targets)
```

Weight tying: reuse embedding matrix as output projection (standard practice, halves parameters).

## Acceptance Criteria

- Given Tiny Shakespeare corpus, when I train for 100 steps, then loss decreases from initial value
- Given a trained model, when I inspect loss over time, then it shows a downward trend
- Given cross_entropy([0.1, 0.2, 0.7], target=2), when I compute it, then loss is -ln(0.7) ≈ 0.357

## Out of Scope

- Learning rate warmup/decay schedules
- Gradient clipping
- Mixed precision (f16/bf16)
- Checkpointing/saving model to disk
- Distributed training
